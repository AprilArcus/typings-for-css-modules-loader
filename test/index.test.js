// @ts-check
/// <reference types="jest" />
const webpack = require("webpack");
const path = require("path");
const memoryfs = require("memory-fs");

beforeEach(() => {
  jest.mock("../src/persist");
});

it("default options", async () => {
  await runTest();

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

it("with sourcemap", async () => {
  await runTest({
    cssLoaderOptions: {
      sourceMap: true
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

it("no modules", async () => {
  await runTest({
    cssLoaderOptions: {
      modules: false
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(0);
});

it("localsConvention asIs", async () => {
  await runTest({
    cssLoaderOptions: {
      localsConvention: "asIs"
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

it("localsConvention camelCase", async () => {
  await runTest({
    cssLoaderOptions: {
      localsConvention: "camelCase"
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

it("with prettier", async () => {
  await runTest({
    options: {
      formatter: "prettier"
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

it("with no formatter", async () => {
  await runTest({
    options: {
      formatter: "none"
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

it("with banner", async () => {
  await runTest({
    options: {
      banner: "// autogenerated by typings-for-css-modules-loader"
    }
  });

  const persisteMock = jest.requireMock("../src/persist");
  expect(persisteMock).toBeCalledTimes(1);
  expect(persisteMock.mock.calls[0][1]).toMatchSnapshot();
});

async function runTest({ options = {}, cssLoaderOptions = {} } = {}) {
  const compiler = webpack({
    entry: path.resolve(__dirname, "./example.css"),
    target: "node",
    module: {
      rules: [
        {
          test: /\.css$/,
          use: [
            {
              loader: require.resolve("../src/index.js"),
              options
            },
            {
              loader: "css-loader",
              options: Object.assign(
                {
                  modules: true
                },
                cssLoaderOptions
              )
            }
          ]
        }
      ]
    },
    mode: "none"
  });

  compiler.outputFileSystem = new memoryfs();

  /** @type {webpack.Stats} */
  const stats = await new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      } else {
        resolve(stats);
      }
    });
  });

  const s = stats.toJson();
  expect(s.errors).toHaveLength(0);
}
