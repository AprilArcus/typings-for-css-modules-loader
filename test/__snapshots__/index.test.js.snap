// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default options 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;
"
`;

exports[`localsConvention asIs 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;
"
`;

exports[`localsConvention camelCase 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    barBaz: string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;
"
`;

exports[`with banner 1`] = `
"// autogenerated by typings-for-css-modules-loader
declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;
"
`;

exports[`with locals export disabled 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss;

export = ExampleCssModule;
"
`;

exports[`with no formatter 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    'bar-baz': string;
    'composed': string;
    'foo': string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;"
`;

exports[`with prettier 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;
"
`;

exports[`with sourcemap 1`] = `
"declare namespace ExampleCssModule {
  export interface IExampleCss {
    \\"bar-baz\\": string;
    composed: string;
    foo: string;
  }
}

declare const ExampleCssModule: ExampleCssModule.IExampleCss & {
  /** WARNING: Only available when \`css-loader\` is used without \`style-loader\` or \`mini-css-extract-plugin\` */
  locals: ExampleCssModule.IExampleCss;
};

export = ExampleCssModule;
"
`;
